File: src/main/scala/SpaceWalk.scala
import com.mongodb.spark._
import com.mongodb.spark.rdd.MongoRDD
import org.bson.Document

val rdd = sc.loadFromMongoDB()

println( rdd.count() )
println ( rdd.first() )
for( doc <- rdd.take( 10 ) ) println( doc )

import com.mongodb.spark.config._

val readConf = ReadConfig( sc )
val readConfig = ReadConfig( Map("collection" -> "eva" ), Some(readConf))
val newRDD = sc.loadFromMongoDB( readConfig = readConfig )

def  breakoutCrew (  document: Document  ): List[(String,Int)]  = {

  var minutes = 0;
  val timeString = document.get( "Duration").asInstanceOf[String]
  if( timeString != null && !timeString.isEmpty ) {
    val time =  document.get( "Duration").asInstanceOf[String].split( ":" )
    minutes = time(0).toInt * 60 + time(1).toInt
  }

  import scala.util.matching.Regex
  val pattern = new Regex("(\\w+\\s\\w+)")
  val names =  pattern findAllIn document.get( "Crew" ).asInstanceOf[String]
  var tuples : List[(String,Int)] = List()
  for ( name <- names ) { tuples = tuples :+ (( name, minutes ) ) }

  return tuples
}

val logs = rdd.flatMap( breakoutCrew ).reduceByKey( (m1: Int, m2: Int) => ( m1 + m2 ) )

logs.foreach( println )

val writeConf = WriteConfig(sc)
val writeConfig = WriteConfig(Map("collection" -> "astronautTotals", "writeConcern.w" -> "majority", "db" -> "nasa"), Some(writeConf))

def mapToDocument( tuple: (String, Int )  ): Document = {
  val doc = new Document();
  doc.put( "name", tuple._1 )
  doc.put( "minutes", tuple._2 )

  return doc
}

logs.map( mapToDocument ).saveToMongoDB( writeConfig )

import org.apache.spark.sql.SQLContext
import com.mongodb.spark.sql._

// load the first dataframe "EVAs"
val evadf = sqlContext.read.mongo()
evadf.printSchema()
evadf.registerTempTable("evas")

// load the 2nd dataframe "astronautTotals"
case class astronautTotal ( name: String, minutes: Integer )
val astronautDF = sqlContext.read.option("collection", "astronautTotals").mongo[astronautTotal]()
astronautDF.printSchema()
astronautDF.registerTempTable("astronautTotals")

sqlContext.sql("SELECT astronautTotals.name, astronautTotals.minutes FROM astronautTotals"  ).show()


sqlContext.sql("SELECT astronautTotals.name, astronautTotals.minutes, evas.Vehicle, evas.Duration FROM " +
  "astronautTotals JOIN evas ON astronautTotals.name LIKE evas.Crew"  ).show()

File: src/main/java/spacewalk/SpaceWalk.java
package spacewalk;

import com.mongodb.spark.api.java.MongoSpark;
import com.mongodb.spark.config.WriteConfig;

import org.bson.Document;
import com.mongodb.spark.rdd.api.java.JavaMongoRDD;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.*;

import scala.Tuple2;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

/**
 * Created by brein on 4/18/2016.
 */
public class SpaceWalk implements Serializable {

    static Pattern pattern = Pattern.compile("(\\w+\\s\\w+)");

    public static void main(String[] args) {

        SparkConf conf = new SparkConf()
                .setMaster("local")
                .setAppName("MongoSparkConnectorTour")
                .set("spark.app.id", "MongoSparkConnectorTour")
                .set("spark.mongodb.input.uri", "mongodb://127.0.0.1/nasa.eva")
                .set("spark.mongodb.output.uri", "mongodb://127.0.0.1/nasa.astronautHours" );


        JavaSparkContext jsc = new JavaSparkContext(conf);
        JavaMongoRDD<Document> rdd = MongoSpark.load(jsc);


        JavaPairRDD<String, Integer> logs = rdd.flatMapToPair(
                new PairFlatMapFunction<Document, String, Integer>
                () {
            @Override
            public Iterable< Tuple2<String, Integer>> call(Document document) throws Exception {
                int minutes = 0;
                String time = (String) document.get("Duration");
                if (!time.isEmpty()) {
                    String[] timeComponents = time.split(":");
                    minutes = (new Integer(timeComponents[0]) * 60);
                    minutes += ( new Integer(timeComponents[1]) );
                }

                String crewString = (String) document.get("Crew");
                Matcher matcher = pattern.matcher(crewString);
                List<String> crew = new ArrayList<String>();

                while (matcher.find()) {
                    crew.add(matcher.group());
                }

                final int finalMinutes = minutes;
                return crew.stream()
                        .map(new Function<String, Tuple2<String, Integer>>() {
                            @Override
                            public Tuple2<String, Integer> apply(String c) {
                                return new Tuple2<>(c, finalMinutes);
                            }
                        })
                        .collect(Collectors.<Tuple2<String, Integer>>toList());
            }
        });

        JavaPairRDD<String, Integer> totalHours =  logs.reduceByKey(
                new Function2<Integer, Integer, Integer>() {
                    @Override
                    public Integer call(Integer v1, Integer v2) throws Exception {
                        return v1 + v2;
                    }
                }
        );

        FlatMapFunction map =  new FlatMapFunction<Tuple2<String,Integer>, Document>(){
            @Override
            public List<Document> call(Tuple2<String, Integer> tuple) throws Exception {
                Document doc = new Document();
                doc.put( "_id", tuple._1 );
                doc.put( "minutes", tuple._2 );
                List <Document> docs = new ArrayList<Document>();
                docs.add( doc );
                return docs;
            }
        };

        JavaRDD<Document>  records  = totalHours.flatMap( map );

        totalHours.foreach(
                new VoidFunction<Tuple2<String, Integer>>() {
                    @Override
                    public void call(Tuple2<String, Integer> tuple ) throws Exception {
                        System.out.println( tuple._1 + " "+ tuple._2  );
                    }
                }
        );

        WriteConfig defaultWriteConfig = WriteConfig.create(jsc);
        Map<String, String> writeOverrides = new HashMap<String, String>();
        writeOverrides.put("collection", "astronautHours");
        writeOverrides.put("databaseName", "nasa");
        writeOverrides.put("writeConcern.w", "majority");
        WriteConfig writeConfig = WriteConfig.create(writeOverrides, defaultWriteConfig);

        MongoSpark.save( records, writeConfig );

    }

}
